File: sdks/js/src/serialize.ts
/**
 * Serialization helpers for WASM C ABI wire format.
 */

function writeU32BE(view: DataView, offset: number, value: number): number {
    view.setUint32(offset, value);
    return offset + 4;
}

function readU32BE(view: DataView, offset: number): [number, number] {
    return [view.getUint32(offset), offset + 4];
}

// ============================================================================
// File array: [num: u32][nameLen: u32][name][dataLen: u32][data]...
// ============================================================================

export function serializeFiles(files: { name: string; data: Uint8Array }[]): Uint8Array {
    const enc = new TextEncoder();
    const total = 4 + files.reduce((n, f) => n + 4 + enc.encode(f.name).length + 4 + f.data.length, 0);
    const buf = new Uint8Array(total);
    const view = new DataView(buf.buffer);
    let off = writeU32BE(view, 0, files.length);

    for (const f of files) {
        const nb = enc.encode(f.name);
        off = writeU32BE(view, off, nb.length);
        buf.set(nb, off); off += nb.length;
        off = writeU32BE(view, off, f.data.length);
        buf.set(f.data, off); off += f.data.length;
    }
    return buf;
}

export function deserializeFiles(data: Uint8Array): { name: string; data: Uint8Array }[] {
    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    const dec = new TextDecoder();
    let [num, off] = readU32BE(view, 0);
    return Array.from({ length: num }, () => {
        let [nl, o] = readU32BE(view, off);
        const name = dec.decode(data.slice(o, o + nl));
        off = o + nl;
        let [dl, o2] = readU32BE(view, off);
        off = o2 + dl;
        return { name, data: data.slice(o2, off) };
    });
}

// ============================================================================
// Chunk array: [num: u32][len: u32][chunk]...
// ============================================================================

export function serializeChunks(chunks: Uint8Array[]): Uint8Array {
    const total = 4 + chunks.reduce((n, c) => n + 4 + c.length, 0);
    const buf = new Uint8Array(total);
    const view = new DataView(buf.buffer);
    let off = writeU32BE(view, 0, chunks.length);

    for (const c of chunks) {
        off = writeU32BE(view, off, c.length);
        buf.set(c, off); off += c.length;
    }
    return buf;
}

export function deserializeChunks(data: Uint8Array): Uint8Array[] {
    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    let [num, off] = readU32BE(view, 0);
    return Array.from({ length: num }, () => {
        let [len, o] = readU32BE(view, off);
        off = o + len;
        return data.slice(o, off);
    });
}

File: sdks/python/src/chithi_sdk/serialize.py
"""Serialization helpers for WASM C ABI wire format."""

import struct


def _serialize_files(files: list[tuple[str, bytes]]) -> bytes:
    """[num:u32 BE][name_len:u32 BE][name][data_len:u32 BE][data]..."""
    parts: list[bytes] = [struct.pack(">I", len(files))]
    for name, data in files:
        name_bytes = name.encode("utf-8")
        parts.append(struct.pack(">I", len(name_bytes)))
        parts.append(name_bytes)
        parts.append(struct.pack(">I", len(data)))
        parts.append(data)
    return b"".join(parts)


def _deserialize_files(data: bytes) -> list[tuple[str, bytes]]:
    """Reverse of _serialize_files."""
    offset = 0
    num = struct.unpack(">I", data[offset:offset + 4])[0]
    offset += 4
    results: list[tuple[str, bytes]] = []
    for _ in range(num):
        name_len = struct.unpack(">I", data[offset:offset + 4])[0]
        offset += 4
        name = data[offset:offset + name_len].decode("utf-8")
        offset += name_len
        data_len = struct.unpack(">I", data[offset:offset + 4])[0]
        offset += 4
        file_data = data[offset:offset + data_len]
        offset += data_len
        results.append((name, file_data))
    return results


def _serialize_chunks(chunks: list[bytes]) -> bytes:
    """[num:u32 BE][len0:u32 BE][chunk0]..."""
    parts: list[bytes] = [struct.pack(">I", len(chunks))]
    for chunk in chunks:
        parts.append(struct.pack(">I", len(chunk)))
        parts.append(chunk)
    return b"".join(parts)


def _deserialize_chunks(data: bytes) -> list[bytes]:
    """Reverse of _serialize_chunks."""
    offset = 0
    num = struct.unpack(">I", data[offset:offset + 4])[0]
    offset += 4
    results: list[bytes] = []
    for _ in range(num):
        length = struct.unpack(">I", data[offset:offset + 4])[0]
        offset += 4
        results.append(data[offset:offset + length])
        offset += length
    return results
